shared/Microsoft.Extensions.TypeNameHelper.Sources/TypeNameHelper.cs

```diff
@@ -36,97 +36,97 @@ public static string GetTypeDisplayName(object item, bool fullName = true)
                 return;
             }
-            if (_builtInTypeNames.ContainsKey(t))
+            if (_builtInTypeNames.ContainsKey(type))
             {
-                sb.Append(_builtInTypeNames[t]);
+                builder.Append(_builtInTypeNames[type]);
             }
             else
             {
-                sb.Append(fullName ? t.FullName : t.Name);
+                builder.Append(fullName ? type.FullName : type.Name);
             }
         }
 
-        private static void ProcessNestedGenericTypes(Type t, StringBuilder sb, bool fullName)
+        private static void ProcessGenericType(StringBuilder builder, Type type, bool fullName)
         {
-            var genericFullName = t.GetGenericTypeDefinition().FullName;
-            var genericSimpleName = t.GetGenericTypeDefinition().Name;
-            var parts = genericFullName.Split('+');
-            var genericArguments = t.GetTypeInfo().GenericTypeArguments;
-            var index = 0;
-            var totalParts = parts.Length;
-            if (totalParts == 1)
+            var genericArguments = type.GetGenericArguments();
+            ProcessNestedGenericType(builder, type, genericArguments, 0, fullName);
+        }
+
+        private static int ProcessNestedGenericType(StringBuilder builder, Type type, Type[] genericArgs, int includedArgsCount, bool fullName)
+        {
+            if (fullName)
+            {
+                if (type.IsNested)
+                {
+                    includedArgsCount = ProcessNestedGenericType(builder, type.DeclaringType, genericArgs, includedArgsCount, fullName);
+                    builder.Append('+');
+                }
+                else
+                {
+                    builder.Append(type.Namespace);
+                    builder.Append('.');
+                }
+            }
+            else
             {
-                var part = parts[0];
-                var num = part.IndexOf('`');
-                if (num == -1) return;
+                // skip generic arguments that belongs to nesting types
+                var nestingType = type.DeclaringType;
+                while (nestingType != null)
+                {
+                    includedArgsCount += nestingType.GetGenericArguments().Length;
+                    nestingType = nestingType.DeclaringType;
+                }
+            }
 
-                var name = part.Substring(0, num);
-                var numberOfGenericTypeArgs = int.Parse(part.Substring(num + 1));
-                sb.Append(fullName ? name : genericSimpleName.Substring(0, genericSimpleName.IndexOf('`')));
-                AppendGenericArguments(genericArguments, index, numberOfGenericTypeArgs, sb, fullName);
-                return;
+            AppendNameOfGenericType(builder, type);
+
+            var numberOfArgs = type.GetGenericArguments().Length;
+            if (numberOfArgs == 0 || includedArgsCount >= genericArgs.Length)
```

Comment:
2nd part of this condition should not be needed. IncludedArgsCount should never exceed genericArgs length. As for equality look for comment on limit

Reply:
It needed in case of non-generic class nested in the generic class.

For example in case of `typeof(Outer<int,string>.Inner)` call to `GetGenericArguments()` will return array of two generic arguments (`int` and `string`) that were already printed for `Outher` class.

So `numberOfArgs` will be `2` and `genericArgs.Length` also will be `2` without second check we will print `Outer<int,string>.Inner<>` that is not valid.


Interestingly: if we use `typeof(Outer<int,string>.Inner).DeclaringType` to get nesting type (that will be `Outer`) we will get open generic type and call to its method `GetGenericArguments()` will return open generic arguments, ex `T1` and `T2`.

Reply:
For extra `<>` it is fine to check on equality. but the check should not require `>=`